### PR TITLE
DRA scheduler: fix queuing hint support

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

d66f8f9 added that "plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods itself".

This applies to DRA because the name of a generated ResourceClaim must be recorded in the pod status before the pod can be scheduled.

#### Does this PR introduce a user-facing change?
```release-note
DRA: when enabling the scheduler queuing hint feature, pods got stuck as unschedulable for a while unnecessarily because recording the name of the generated ResourceClaim did not trigger scheduling.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
- [KEP]: https://github.com/kubernetes/enhancements/issues/4247
```

/assign @sanposhiho

### CODE REVIEW

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:403]: In order to reduce the number of subscribing events, we avoid using `Pod/Update` and instead use [specific events `Pod/UpdatePodXXXX`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L69-L76).

So, I think you need to create a new one like `Pod/UpdatePodResourceClaims`.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (googs1025) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:405]: Does DeviceClass need the same operation? If so, will you submit it in this PR? Or can I submit another PR to complete this task and add some unit tests? @pohly

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:405]: I think DeviceClass is fine the way it is. It's very unlikely to change, so there is no risk of having to process too many events. That is the concern for pods.

Comment (googs1025) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:405]: got it

File: pkg/scheduler/framework/events.go
Comment (toVersus) [inline pkg/scheduler/framework/events.go:231]: Since PodResourceClaimStatus doesn't have fields that are updated later, like state or timestamps, and entries added to PodResourceClaimStatus are never removed, does that mean we only need to compare the length of the slice? I saw the comment in another part of the code that caught my eye.
https://github.com/kubernetes/kubernetes/blob/52095a8b7b9b75d67a3882a21a6647e4f90ade48/plugin/pkg/auth/authorizer/node/graph_populator.go#L114-L116

Comment (pohly) [reply pkg/scheduler/framework/events.go:231]: If we want to be pedantic, then yes, we would have to do a deep equal. As I decided to be pedantic elsewhere, let's to the same here :grin: 

Thanks for reminding me.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:527]: Can you change the level to 5, if you don't have a strong preference. (Other QHints have level5 only

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:523]: log level?

File: pkg/scheduler/framework/events.go
Comment (sanposhiho) [reply pkg/scheduler/framework/events.go:231]: Can you leave the similar comment about why we only need check len?

**Review by sanposhiho — COMMENTED** [review]
only nits

File: pkg/scheduler/framework/events.go
Comment (pohly) [reply pkg/scheduler/framework/events.go:231]: I'm now using a common helper function to check not only the length, but also the content.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:527]: Changed.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:523]: Oops, yes. This was copy-and-pasted from above where it uses `loggerV` and then I forgot to add `V(6)` back.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Also in other, older code.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (sanposhiho) [inline staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:32]: Could statusA and statusB have the same items, but with different orders?

**Review by sanposhiho — APPROVED** [review]
/approve

Leave `/lgtm` to someone else, @macsko or @alculquicondor.
/cc @macsko @alculquicondor

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:32]: In theory, yes. In practice that should never happen, so adding a `slices.Sort` here just adds performance overhead in the common case that doesn't pay off.

**Comment by pohly** [comment]
> pull-kubernetes-kind-dra: cluster supports count/resourceclaims.resource.k8s.io ResourceQuota 

A flake: the apiserver should prevent adding a second ResourceClaim, but does not always do so. A timing issue?!

>  pull-kubernetes-dependencies

@liggitt: this refactoring will need your approval because I am moving some helper function out of `authorizer/node` into a shared package.

But before this can be merged: what should I do about the [unwanted dependency on go-fuzz](https://github.com/kubernetes/kubernetes/pull/125487)? Are there alternatives that can be used for generating random test cases?

As I said in staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go: I could manually write test cases, but then the test won't catch when someone forgets to extend the code when adding some new field.

**Comment by alculquicondor** [comment]
cc @dom4ha

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (liggitt) [inline staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:25]: Does this treat `nil` and `[]` as equal? That's fine for semantic equivalence (and was fine for the authorization use), but could mess with callers who expect reflect.DeepEqual to agree with the results here.

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:25]: Semantic equivalence is sufficient.

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:25]: I'm comparing against reflect.DeepEqual, but apparently don't hit that particular kind of "not quite equal" when using go-fuzz.

Comment (liggitt) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:25]: > Semantic equivalence is sufficient.

I agree, mostly something we should make the naming / documentation clear on

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go]: Renamed to just `PodStatusEqual` and updated the documentation.

I stopped using go-fuzz and replaced it with manually written testcases. That should be okay, no changes to `corev1.PodResourceClaimStatus` are expected. I'm still wondering what the replacement will be in other places were we use it.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (macsko) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: Don't we need any tests for that?

Comment (googs1025) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: hey, @macsko @pohly I am working on this PR, which seems to be related to this. If necessary, I can add this part of the test case to this PR :)
https://github.com/kubernetes/kubernetes/pull/127523

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: @macsko: yes, this needs tests. Is it okay if @googs1025 adds them as a follow-up?

Comment (macsko) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: Sure, can be this way

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go]: /assign @liggitt 

For final approval.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (dom4ha) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: The alternative would be to just check whether the generated resource claims has shrunk down and enable scheduling, but I suspect this was a conscious decision.

Comment (dom4ha) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: Actually shrunk down or changed, as in theory they claims can change as well. So not sure what is the most common case and whether the additional complication is worth the effort, but changes that are append only, won't infer the scheduling I think.

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: This hint function only gets called if the code over in `events.go` has already determined that the pod's generated claims changed, so there is no need to check that again.

What the code here checks is "are all required claims generated".

Comment (dom4ha) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: > This hint function only gets called if the code over in `events.go` has already determined that the pod's generated claims changed, so there is no need to check that again.

For instance instead of introducing a generic change event UpdatePodGeneratedResourceClaim, the alternative one could be scoped down (emitted) to changes that shank down the Pod's claims (if any of them disappeared - PodGeneratedResourceClaimEliminated) as this is the only case worth requeuing.

Unfortunately I don't know how often it happens that new ResourceClaims just grow, as this is the only case this approach could eliminate (isSchedulableAfterPodChange would not be called).

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: > the alternative one could be scoped down (emitted) to changes that shank down the Pod's claims (if any of them disappeared - PodGeneratedResourceClaimEliminated) as this is the only case worth requeuing

It's the opposite: *adding* entries may make a pod schedulable because the ResourceClaim that it needs got added.

In practice, entries only get added, never removed nor updated. I don't think it's worth making the code more complex for situations that don't occur.

Comment (dom4ha) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:525]: Ah right, claims gets fulfilled over time. There's indeed no place for optimization.
Sorry for the dumb question, just ramping up and connecting dots slowly :)

**Comment by dom4ha** [comment]
LGTM

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: I was going to ask whether adding a specific event is rather overkill, given that there is only one plugin that cares about it, and it is alpha.
In other words, we could hide the complexity inside the hint, as opposed to make it part of the framework.

But, it's all about prons and cons. I'll leave the decision to @sanposhiho

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: IIRC, the idea of fine-grained events is more important when there are no scheduling hints. But DRA is as close to graduating to beta as scheduling hints are, so maybe it's worth not adding extra complexity in the framework.

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Oh right, having fine grained events is helpful to reduce the number of stored inflight events, which affects scheduling throughput even if DRA is not used in the cluster. That sounds worse for the average user. Ok, let's keep the fine-grained event :)

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: This is not how we do things. And even more so when we have gotten rid of "test freeze", meaning that code needs to merge with all appropriate coverage.

I've let similar situations slide in the past, but IIRC, DRA is targeting beta now, so we can't be as forgiving as we were in previous iterations of DRA.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (alculquicondor) [inline staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:49]: I think you can use `ptr.Equal`

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (johnbelamaric) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: @alculquicondor +1, we need to be sure we merge code and tests together from now on

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Yes, I think that was the reason why @sanposhiho suggested it.

If DRA is disabled, the new event isn't getting generated. But if it's enabled (not on by default while beta!) and merely not used (no ResourceClaims), then it will add some overhead.

File: test/integration/scheduler_perf/.gitignore
Comment (pohly) [reply test/integration/scheduler_perf/.gitignore:2]: Done.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:49]: Indeed, thanks for the tip.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:507]: Fair enough. I've added unit tests.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1520]: we don't check these objects in the function, do we?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1513]: add a case for different UID

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1554]:
```suggestion
			gotHint, err := testCtx.p.isSchedulableAfterPodChange(logger, tc.pod, nil, tc.obj)
```

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1512]:
```suggestion
		wantHint framework.QueueingHint
		wantErr  bool
```

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go
Comment (alculquicondor) [inline staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: In SIG scheduling we prefer not to use assertion libraries. Please use `t.Error`

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1560]: Avoid assertion libraries and write meaningful errors for `t.Error`

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go
Comment (alculquicondor) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1520]: We do, indirectly in https://github.com/kubernetes/kubernetes/blob/3d6c5b2e98afaaae1d17107e2d3d709c726be49d/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L683-L686

As part of my work on https://github.com/kubernetes/kubernetes/pull/127277 I checked whether removing it would make a difference, but couldn't measure any, so I decided to keep the `Get` as an additional check.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1513]: Added.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: As SIG Testing tech lead I don't agree with those guidelines. I accept it for code owned by SIG scheduling, but this here is a general-purpose helper package.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1560]: Okay, changed it here.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go
Comment (liggitt) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: Testify keeps us tied to a yaml library we want to drop eventually (https://github.com/stretchr/testify/blob/master/go.mod#L11), it would be vaguely nice to not add more deps to it

Comment (liggitt) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: (But not a blocker, certainly)

**Comment by liggitt** [comment]
/approve
For authorizer change

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go]: That's not what is used elsewhere, for example:
https://github.com/kubernetes/kubernetes/blob/3d6c5b2e98afaaae1d17107e2d3d709c726be49d/pkg/scheduler/framework/plugins/nodeports/node_ports_test.go#L305-L358

But I don't care, so I changed it in `dynamicresources_test.go` also in the other functions.

File: staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: https://github.com/stretchr/testify/pull/1579 added build tags which avoid that dependency. I'll have a look whether we can leverage that.

Comment (liggitt) [reply staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:23]: We can't :-(

go dependency and vendoring don't consider build tags since they have to resolve ahead of when builds are done and be valid for whatever builds *could* be done

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**
